include/cantera/numerics/AdaptivePreconditioner.h

```diff
@@ -0,0 +1,212 @@
+
+    //! Function used to return semi-analytical jacobian matrix
+    Eigen::SparseMatrix<double> getJacobian(){
+        Eigen::SparseMatrix<double> jacobian(m_dimensions[0], m_dimensions[1]);
+        jacobian.setFromTriplets(m_jac_trips.begin(), m_jac_trips.end());
+        return jacobian;
+    };
+
+    //! Use this function to get derivative settings
+    void getPreconditionerDerivativeSettings(AnyMap& settings){
+        settings = m_settings;
+    }
+
+    //! Use this function to get the threshold value for setting
+    //! elements
+    double getThreshold(){return m_threshold;};
+
+    //! Use this function to get the pertubation constant
+    double getPerturbationConst(){return m_perturb;};
+
+    //! Use this function to get a strictly positive composition
+    void getStrictlyPositiveComposition(size_t vlen, double* in, double* out){
+        for (size_t i = 0; i < vlen; i++)
+        {
+            out[i] = std::max(in[i], m_atol);
+        }
+    };
+
+    //! Use this function to set derivative settings
+    void setPreconditionerDerivativeSettings(AnyMap& settings){
+        m_settings = settings;
+    }
+
+    //! Use this function to set the threshold value to compare elements
+    //! against
+    //! @param threshold double value used in setting by threshold
+    void setThreshold(double threshold){
+        m_threshold = threshold;
+        m_prune_precon = (threshold <= 0) ? false : true;
+    };
+
+    //! Use this function to set drop tolerance for ILUT
+    //! @param droptol double value used in setting solver drop tolerance
+    void setDropTolILUT(double droptol = 1e-10){
+        m_solver.setDroptol(droptol);
+        };
+
+    //! Use this function to set the fill factor for ILUT
+    void setFillFactorILUT(int fillfactor = -1){
+        int newfillfactor = (fillfactor < 0) ? m_dimensions[0]/4 : fillfactor;
+        m_solver.setFillfactor(newfillfactor);
+    }
+
+    //! Use this function to set the perturbation constant used in
+    //! finite difference calculations.
+    //! @param perturb the new pertubation constant
+    void setPerturbationConst(double perturb){
+        m_perturb = perturb;
+    };
+
+    //! Overloading of the == operator to compare values strictly inside
+    //! preconditioner matrix
+    //! @param externalPrecon == comparison with this object
+    bool operator== (const AdaptivePreconditioner &externalPrecon);
+    //! Overloading of the = operator to copy one preconditioner to
+    //! another
+    //! @param externalPrecon the preconditioner becoming this object
+    void operator= (const AdaptivePreconditioner &externalPrecon);
+
+    //! Overloading of the () operator to assign values to the jacobian
+    //! this function does not assume that index is index map
+    //! @param row row index of jacobian
+    //! @param col column index of jacobian
+    //! @param value to place in jacobian vector
+    void operator() (size_t row, size_t col, double value);
```

Comment:
@speth The overloading was done this way because I am using the Eigen recommended way of building sparse matrices (a vector of Triplets) which to my knowledge cannot be mutated once it is pushed into the vector.